### PR TITLE
Include Entry Assembly File Location when loading candidate NLog.config

### LINKED ISSUE #1342: Config file not found when running via .NET Core
We have a DNX application that we run through Visual Studio or via the command line as in 

```
dnx.exe --project "[path to my project.json]" run
```

There is an _NLog.config_ file in the project's folder, but NLog does not find it. This is because NLog is looking in the `BaseDirectory` of the current app domain, and under this environment that directory is the directory of DNX (where dnx.exe resides). In this scenario the correct directory is given by `Directory.GetCurrentDirectory()`.

**Type**:
- Bug

**NLog version**: 4.2.3 (current build also)

**Platform**: DNX 1.0.0-RC1-Final

### PR DESCRIPTION
**Pull request by snakefoot**
Improve the NetCore experience where the BaseDirectory is often different from the bin-location (But the Bin-location is where nlog.config is copied to by default)

Also improve the logic for recognizing identical paths (with or without ending directory-separator)

Think this will resolve #1342 (Atleast for NetStandard1.5 and NetStandard2.0)

### CODE REVIEW

File: src/NLog/Internal/AssemblyHelpers.cs [hunk 1]
```diff
@@ -107,5 +107,81 @@ public static Assembly LoadFromName(string assemblyName)
 #endif
         }
 
+#if !SILVERLIGHT && !NETSTANDARD1_3
+        public static string GetAssemblyFileLocation(Assembly assembly)
```
Comment (304NotModified) [inline src/NLog/Internal/AssemblyHelpers.cs]: is this just a move or are there some changes here?

File: src/NLog/LogFactory.cs [hunk 2]
```diff
@@ -1092,15 +1093,51 @@ public void ResetCandidateConfigFilePath()
         private static IEnumerable<string> GetDefaultCandidateConfigFilePaths()
         {
             // NLog.config from application directory
-            if (CurrentAppDomain?.BaseDirectory != null)
+            string baseDirectory = PathHelpers.TrimDirectorySeparators(CurrentAppDomain?.BaseDirectory);
+            if (!string.IsNullOrEmpty(baseDirectory))
+                yield return Path.Combine(baseDirectory, "NLog.config");
+
+            bool platformFileSystemCaseInsensitive = PlatformDetector.IsWin32;
+            if (!platformFileSystemCaseInsensitive && !string.IsNullOrEmpty(baseDirectory))
```
Comment (304NotModified) [inline src/NLog/LogFactory.cs]: PS: windows could be case sensitive, but highly unlikely ;)

File: src/NLog/LogFactory.cs [hunk 3]
```diff
@@ -1092,15 +1093,51 @@ public void ResetCandidateConfigFilePath()
         private static IEnumerable<string> GetDefaultCandidateConfigFilePaths()
         {
             // NLog.config from application directory
-            if (CurrentAppDomain?.BaseDirectory != null)
+            string baseDirectory = PathHelpers.TrimDirectorySeparators(CurrentAppDomain?.BaseDirectory);
+            if (!string.IsNullOrEmpty(baseDirectory))
```
Comment (304NotModified) [inline src/NLog/LogFactory.cs]: please move the check `string.IsNullOrEmpty(baseDirectory)` on top of both `if`s - it's now duplicated

(so we have 3 ifs, 1 top and 2 nested)

Comment (snakefoot) [reply src/NLog/LogFactory.cs]: Well I like to skip the `PlatformDetector.IsWin32` when possible.

File: src/NLog/Internal/AssemblyHelpers.cs (hunk 1 above)
Comment (snakefoot) [reply src/NLog/Internal/AssemblyHelpers.cs]: It is a move and some change in the internallogger output

File: src/NLog/LogFactory.cs (hunk 2 above)
Comment (snakefoot) [reply src/NLog/LogFactory.cs]: Would cause most applications on Windows to fail badly, which is why I skip the extra file-checks

**Comment by snakefoot** [comment]
Removed the usage of NLog-assembly-location for NLog.config loading, as NetCore will most likely reference NLog.dll in nuget-folder.

File: src/NLog/LogFactory.cs (hunk 3 above)
Comment (304NotModified) [reply src/NLog/LogFactory.cs]: I'm a bit confused, currently it tests the `PlatformDetector.IsWin32` before the baseDirectory check, so it isn't skipped when possible?

To be clear, the proposal:


```c#

if (!string.IsNullOrEmpty(baseDirectory))
{
    yield return Path.Combine(baseDirectory, "NLog.config");
    bool platformFileSystemCaseInsensitive = PlatformDetector.IsWin32;
    if (!platformFileSystemCaseInsensitive)
    {
         yield return Path.Combine(baseDirectory, "nlog.config");
    }

}

```

`baseDirectory` was checked before, so no reason to check it again IMO

Comment (snakefoot) [reply src/NLog/LogFactory.cs]: PlatformDetector.IsWin32 is only called once in the current implementation. Your suggestions requires multiple lookups.

Comment (304NotModified) [reply src/NLog/LogFactory.cs]: OK, Fair enough

**Pull request merged**